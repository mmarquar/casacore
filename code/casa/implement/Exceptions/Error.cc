//# Error.cc: Base class for all AIPS++ errors (templated classes)
//# Copyright (C) 1993,1994,1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Exceptions/Error.h>
#include <iostream.h>


//#  This template is for index errors  where the Index can be any type. It
//#  is derived from the class IndexError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(indexError);
rtti_imp_mbrf_a1(indexError);

template<class t>
indexError<t>::indexError(t oI, const Char *str) : oIndex(oI), IndexError(str)
{}
template<class t>
indexError<t>::indexError(t oI, const String &str) : oIndex(oI), IndexError(str)
{}

template<class t>
indexError<t>::indexError(ExcpError *excp) : IndexError(excp)
{
    indexError<t> *tmp;
    PCAST(tmp,indexError<t>,excp);
    if (tmp) {
        oIndex = (*tmp).oIndex;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
indexError<t>::~indexError()
    {}

//#  This template is for duplicate key errors  where the Key can be any type. It
//#  is derived from the class DuplError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(duplError);
rtti_imp_mbrf_a1(duplError);

template<class t>
duplError<t>::duplError(ExcpError *excp) : DuplError(excp)
{
    duplError<t> *tmp;
    PCAST(tmp,duplError<t>,excp);
    if (tmp) {
        oKey = (*tmp).oKey;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
duplError<t>::~duplError()
    {}


template<class t>
duplError<t>::duplError(t oI, const Char *str) : oKey(oI), DuplError(str)
{}
template<class t>
duplError<t>::duplError(t oI, const String &str) : oKey(oI), DuplError(str)
{}

