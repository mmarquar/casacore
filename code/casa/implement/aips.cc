//# aips.cc: Global initialization for namespace management, standard types, etc.
//# Copyright (C) 1993,1994,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/aips.h>
#include <stdio.h>
#include <stdlib.h>

#if defined(AIPS_DEBUG)
Bool aips_debug_on = True;
#else
Bool aips_debug_on = False;
#endif

static size_t total = 0;
const size_t &aipsalloc(total);

// Replace global operator new/delete unless the user says to do otherwise.
// We do this to keep track of memory use.
#if !defined(AIPS_STDNEW)
#include <new.h>

// Assume everything aligns on 8 byte boundaries. No portable way to determine
// this. We store the allocation size "before" the new'd pointer in an area this
// long
const int offset = 8;

// When we have modern compilers we should throw an exception if the new_handler
// fails.

void *operator new (size_t size)
{
    char *ptr;
    while ((ptr = (char *)malloc(size+offset)) == 0) {
	new_handler handler;
	handler = set_new_handler(0);
	set_new_handler(handler); // Put it back before trying it
	if (handler) {
	    (*handler)();
	} else {
	    return 0;
	}
    }
    *((size_t *)ptr) = size;
    if (sizeof(size_t)*2 <= offset) {
	// Write it twice if it will fit as a sanity check.
	*((size_t *)ptr+1) = size;
    }
    total += size;
    return ptr+offset;
}

void operator delete (void *ptr)
{
    if (ptr) {
	char *alias = (char *)ptr;
	alias -= offset;
	size_t size = *((size_t*)alias);
	if (sizeof(size_t)*2 <= offset) {
	    if (size != *((size_t*)alias+1)) {
		fprintf(stderr, "ERROR: delete'ing memory not allocated "
			"with new!\n");
		exit(999);
	    }
	}
	total -= size;
	free(alias);
    }
}

//------ Just copies of the above with [] added

void *operator new[] (size_t size)
{
    char *ptr;
    while ((ptr = (char *)malloc(size+offset)) == 0) {
	new_handler handler;
	handler = set_new_handler(0);
	set_new_handler(handler); // Put it back before trying it
	if (handler) {
	    (*handler)();
	} else {
	    return 0;
	}
    }
    *((size_t *)ptr) = size;
    if (sizeof(size_t)*2 <= offset) {
	// Write it twice if it will fit as a sanity check.
	*((size_t *)ptr+1) = size;
    }
    total += size;
    return ptr+offset;
}

void operator delete[] (void *ptr)
{
    if (ptr) {
	char *alias = (char *)ptr;
	alias -= offset;
	size_t size = *((size_t*)alias);
	if (sizeof(size_t)*2 <= offset) {
	    if (size != *((size_t*)alias+1)) {
		fprintf(stderr, "ERROR: delete'ing memory not allocated "
			"with new!\n");
		exit(999);
	    }
	}
	total -= size;
	free(alias);
    }
}

#endif
