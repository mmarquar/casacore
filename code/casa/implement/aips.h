//# aips.h: Global initialization for namespace management, standard types, etc.
//# Copyright (C) 1993-1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_ENTER_)
#include <aips/aips_enter.h>
#endif

#if !defined(AIPS_AIPS_H)
#define AIPS_AIPS_H

#if defined(_AIX)
#pragma implementation ("aips.cc")
#endif

// At least the centerline compiler doesn't know what to do with volatile.
// This may be more generally true.

#if defined(OBJECTCENTER)
#define volatile
                    // nothing
#endif

// Including aips.h should guarantee that AIPS++ namespace management is
// switched on - it may have been explicitly switched off beforehand by
// aips_exit.h.

// Defined the "aips_" macro which is used for name space management, and
// the "aips_name2" macro which is used to join two tokens.

#if defined(__STDC__) || defined(__ANSI_CPP__) || defined(__hpux)
#define aips_(x)  aips##x
#define aips_name2(a,b) a##b
#else
#define aips_(x)  aips/**/x
#define aips_name2(a,b) a/**/b
#endif

// This section contains the various standard types used by AIPS++.

enum Bool { False = 0, True = 1 };

typedef char Char;
typedef unsigned char uChar;
typedef short Short;
typedef unsigned short uShort;
typedef int Int;
typedef unsigned int uInt;
typedef long Long;
typedef unsigned long uLong;
typedef float Float;
typedef double Double;
#if defined (AIPS_IRIX) && !defined(__GNUG__)
typedef double lDouble;
#else
typedef long double lDouble;
#endif

#if defined(__GNUG__)
// Needed for g++ 2.7.2. Hopefully we can delete these at some point.
// <group>
inline const Bool &at_cc(const Bool &val) { return(val); };
inline const Char &at_cc(const Char &val) { return(val); };
inline const uChar &at_cc(const uChar &val) { return(val); };
inline const Short &at_cc(const Short &val) { return(val); };
inline const uShort &at_cc(const uShort &val) { return(val); };
inline const Int &at_cc(const Int &val) { return(val); };
inline const uInt &at_cc(const uInt &val) { return(val); };
inline const Long &at_cc(const Long &val) { return(val); };
inline const uLong &at_cc(const uLong &val) { return(val); };
inline const Float &at_cc(const Float &val) { return(val); };
inline const Double &at_cc(const Double &val) { return(val); };
inline const lDouble &at_cc(const lDouble &val) { return(val); };

inline Bool &at_c(Bool &val) { return(val); };
inline Char &at_c(Char &val) { return(val); };
inline uChar &at_c(uChar &val) { return(val); };
inline Short &at_c(Short &val) { return(val); };
inline uShort &at_c(uShort &val) { return(val); };
inline Int &at_c(Int &val) { return(val); };
inline uInt &at_c(uInt &val) { return(val); };
inline Long &at_c(Long &val) { return(val); };
inline uLong &at_c(uLong &val) { return(val); };
inline Float &at_c(Float &val) { return(val); };
inline Double &at_c(Double &val) { return(val); };
inline lDouble &at_c(lDouble &val) { return(val); };
// </group>
#else    /* !__GNUG__ */
#define at_c(X) X
#define at_cc(X) X
#endif   /* __GNUG__ */

// This is an inline guaranteed-correct conversion function which converts
// to Bool any type which has meaning in a C++ logical expression.
template <class T>
inline
Bool ToBool (T val)
{
    return (val ? True : False);
}

// If AIPS_DEBUG is not defined, then the symbol expands to (0) which in an
// if should be removed by the dead code eliminator of any optimizer; thus
// using this in your code should have no performance penalty in the normal
// case. If compiled with AIPS_DEBUG, then aips_debug is (defined to )
// a global boolean variable (so it can be turned on and off in a debugger) 
// which is initialized to True.

extern Bool aips_debug_on;

#if !defined(AIPS_DEBUG)
#define aips_debug (0)
#else
// The reason that we just don't make this a variable here is so that
// we can link against libraries compiled with or without AIPS_DEBUG
// without having any missing symbols.
#define aips_debug aips_debug_on
#endif

// This is just syntactic sugar to tell the namespace management that a
// particular class is not under control of namespace management. The only
// place this will normally be needed is forward declarations of imported
// classes, e.g. from InterViews.
#define imported /* Nothing */

// These are new keywords in the upcoming C++ standard.
// They are defined to an empty string, so they can be used until
// all compilers interpret them.
// Especially the keyword explicit may give useful information.
// It can be given at a one-argument constructor telling the compiler
// that it should not be used for automatic conversions.
// E.g.    explicit IPosition (uInt);
#define explicit /* Nothing */
#define typename /* Nothing */

#endif
