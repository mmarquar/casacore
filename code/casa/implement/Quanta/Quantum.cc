//# Quantum.cc: class to manipulate physical, dimensioned quantities
//# Copyright (C) 1994,1995,1996,1997,1998,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <aips/Mathematics/Complex.h>
#include <aips/Mathematics/Constants.h>
#include <aips/Lattices/IPosition.h>
#include <aips/Arrays/Array.h>
#include <aips/Arrays/ArrayMath.h>
#include <aips/Arrays/ArrayLogical.h>
#include <aips/Arrays/Vector.h>
#include <aips/Arrays/Matrix.h>
#include <aips/Arrays/Cube.h>
#include <aips/Exceptions/Error.h>
#include <aips/Quanta/Quantum.h>
#include <aips/Quanta/MUString.h>
#include <aips/Quanta/MVAngle.h>
#include <aips/Quanta/MVTime.h>
#include <aips/RTTI/Register.h>

template <class Qtype>
Quantum<Qtype>::Quantum() :
    QBase() { qVal = Qtype();}

template <class Qtype>
Quantum<Qtype>::Quantum(const Quantum<Qtype> &other) :
    QBase(other), qVal(other.qVal) {}

template <class Qtype>
Quantum<Qtype>::Quantum(const Qtype &factor) : QBase(),
    qVal(factor) {}

template <class Qtype>
Quantum<Qtype>::Quantum(const Qtype &factor, const Unit &s) :
  QBase(s),
  qVal(factor) {}

template <class Qtype>
Quantum<Qtype>::Quantum(const Qtype &factor, const QBase &other) :
  QBase(other),
  qVal(factor) {}

template <class Qtype>
Quantum<Qtype>::~Quantum() {}

//# Quantum operators

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator=(const Quantum<Qtype> &other) {
    if (this != &other) {
      qVal=other.qVal;
      qUnit=other.qUnit;
    }
    return *this;
}

template <class Qtype>
const Quantum<Qtype> &Quantum<Qtype>::operator+() const{
    return *this;
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::operator-() const{
    Quantum<Qtype> loc;
    loc.qVal = -qVal;
    loc.qUnit = qUnit;
    return loc;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator+=(const Quantum<Qtype> &other) {
    if (qUnit.getValue() != other.qUnit.getValue()) {
	throw (AipsError("Quantum::operator+ unequal units '" +
			 qUnit.getName() + ", '" + 
			 other.qUnit.getName() + "'"));
    } else {
        Qtype tmp = other.getValue(qUnit);
	qVal += (tmp);
    }
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator+=(const Qtype &other) {
    qVal += other;
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator-=(const Quantum<Qtype> &other) {
    if (qUnit.getValue() != other.qUnit.getValue()) {
	throw (AipsError("Quantum::operator- unequal units '" +
			 qUnit.getName() + ", '" + 
			 other.qUnit.getName() + "'"));
    } else {
        Qtype tmp = other.getValue(qUnit);
	qVal -= (tmp);
    }
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator-=(const Qtype &other) {
    qVal -= (other);
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator*=(const Quantum<Qtype> &other) {
    qVal *= (other.qVal); 
    if (!(other.qUnit.getName().empty())) {
	if (qUnit.getName().empty()) {
	    qUnit = other.qUnit;
	} else {
	    qUnit = Unit(qUnit.getName() + ("." + other.qUnit.getName()));
	}
    }
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator*=(const Qtype &other) {
    qVal *= (other);
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator/=(const Quantum<Qtype> &other) {
    qVal /= (other.qVal);
    if (!(other.qUnit.getName().empty())) {
	if (qUnit.getName().empty()) {
	    qUnit = Unit(String("(") + other.qUnit.getName() +
			 String(")-1"));
	} else {
	    qUnit = Unit(qUnit.getName() + ("/(" + 
					    other.qUnit.getName() + ")"));
	}
    }
    return *this;
}

template <class Qtype>
Quantum<Qtype> &Quantum<Qtype>::operator/=(const Qtype &other) {
    qVal /= (other);
    return *this;
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::operator+(const Quantum<Qtype> &other) const{
    Quantum<Qtype> loc; loc = *this;
    loc += other;
    return loc;
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::operator-(const Quantum<Qtype> &other) const{
    Quantum<Qtype> loc; loc = *this;
    loc -= other;
    return loc;
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::operator*(const Quantum<Qtype> &other) const{
    Quantum<Qtype> loc; loc = *this; 
    loc *= other;
    return loc;
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::operator/(const Quantum<Qtype> &other) const{
    Quantum<Qtype> loc; loc = *this;
    loc /= other;
    return loc;
}

template <class Qtype>
void  Quantum<Qtype>::print(ostream &os) const {
    os << qVal << " " << qUnit.getName();
}

//# Quantum general member functions

template <class Qtype>
const Qtype &Quantum<Qtype>::getValue() const {
    return qVal;
}

template <class Qtype>
Qtype & Quantum<Qtype>::getValue() {
  return qVal;
}

template <class Qtype>
Qtype Quantum<Qtype>::getValue(const Unit &other) const {
    Double d1 = other.getValue().getFac() /
	qUnit.getValue().getFac();	// SUN native overloading problems
    if (qUnit.getValue() == UnitVal::ANGLE) {
      if (other.getValue() == UnitVal::TIME)
	d1 *= C::circle/C::day;
    } else if (qUnit.getValue() == UnitVal::TIME) {
      if (other.getValue() == UnitVal::ANGLE)
	d1 *= C::day/C::circle;
    };
    return (Qtype)(qVal/d1);
}

template <class Qtype>
Qtype Quantum<Qtype>::getBaseValue() const {
    return (Qtype)(qVal * qUnit.getValue().getFac());
}

template <class Qtype>
const Unit &Quantum<Qtype>::getFullUnit() const {
    return qUnit;
}

template <class Qtype>
void Quantum<Qtype>::scale(const Qtype &factor) {
    qVal *= (factor);
}

template <class Qtype>
void Quantum<Qtype>::setValue(const Qtype &val) {
    qVal = val;
}

template <class Qtype>
Bool Quantum<Qtype>::read(Quantity &res, MUString &in) {
  Double val0 = 0.0;
  String unit = "";
  res = Quantity();
  UnitVal uv;
  in.push();
  if (!in.eos()) {
    if (MVAngle::read(res, in) || MVTime::read(res, in)) {
      val0 = res.getValue();
      unit = res.getUnit();
    } else {
      val0 = in.getDouble();
      unit = in.get();
      // Check if valid unit specified
      if (!UnitVal::check(unit, uv)) {
	in.pop(); return False;
      };
    };
  };
  //
  // The next statement is necessary once the read return arg is templated
  //  Qtype tmp = (Qtype)((res.getValue()) + val0)
  res.setValue(val0);
  res.setUnit(unit);
  in.unpush(); return True; 
}

template <class Qtype>
Bool Quantum<Qtype>::read(Quantity &res, const String &in) {
  MUString tmp(in);		// Pointed non-const String
  return Quantum<Qtype>::read(res, tmp);
}

template <class Qtype>
Bool Quantum<Qtype>::check(const UnitVal &uv) const {
    return ( (qUnit.getValue() == uv) ? True : False); 
}

template <class Qtype>
void Quantum<Qtype>::assert(const UnitVal &uv) const {
    if (qUnit.getValue() != uv) {
	throw(AipsError("Quantum::assert non-conforming unit type '" +
			getUnit() + "'"));
    }
}

template <class Qtype>
void Quantum<Qtype>::convert() {
    this->convert(Unit());
}

template <class Qtype>
void Quantum<Qtype>::convert(const Unit &s) {
    if (qUnit.getValue() == s.getValue()) {
      // To suppress some warnings, next statement not used
      //	qVal *= (qUnit.getValue().getFac()/s.getValue().getFac());
      qVal = (Qtype) (qVal * 
		      qUnit.getValue().getFac()/s.getValue().getFac());
      qUnit = s;
    } else {
      if (qUnit.getValue() == UnitVal::ANGLE && 
	  s.getValue() == UnitVal::TIME) {
	qVal = (Qtype) (qVal *
			qUnit.getValue().getFac()/
			s.getValue().getFac() * C::day/C::circle);
	qUnit = s;
      } else if (qUnit.getValue() == UnitVal::TIME &&
		 s.getValue() == UnitVal::ANGLE) {
	qVal = (Qtype) (qVal *
			qUnit.getValue().getFac()/
			s.getValue().getFac() * C::circle/C::day);
	qUnit = s;
      } else {
	qUnit.setValue(qUnit.getValue() / s.getValue());
	ostrstream oss;
	oss << qUnit.getValue().getDim();
	// Suppress (gcc) warnings:
	qVal = (Qtype) (qVal * 
			      qUnit.getValue().getFac());
	if (s.empty()) {
	  qUnit = String(oss);
	} else {
	  qUnit = Unit(s.getName() + '.' + String(oss).after(0));
	};
      };
    };
}

template <class Qtype>
void Quantum<Qtype>::convert(const Quantum<Qtype> &other) {
    this->convert(other.qUnit);
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::get() const {
    return get(Unit());
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::get(const Unit &s) const {
    Quantum<Qtype> res = *this;
    res.convert(s); return res;
}

template <class Qtype>
Quantum<Qtype> Quantum<Qtype>::get(const Quantum<Qtype> &other) const{
    return get(other.qUnit);
}

template <class Qtype>
QBase *Quantum<Qtype>::clone() const {
  return (new Quantum<Qtype>(*this));
}

template <class Qtype>
uInt Quantum<Qtype>::type() const {
  return Register((Quantum<Qtype> *)0);
}

template <class Qtype>
uInt Quantum<Qtype>::myType() {
  return Register((Quantum<Qtype> *)0);
}

