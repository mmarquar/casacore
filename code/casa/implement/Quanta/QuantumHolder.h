//# QuantumHolder.h: A holder for Quantities to enable record conversions
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_QUANTUMHOLDER_H)
#define AIPS_QUANTUMHOLDER_H

//# Includes
#include <aips/aips.h>
#include <aips/Utilities/PtrHolder.h>
#include <trial/Utilities/RecordTransformable.h>

//# Forward Declarations
template <class Qtype> class Quantum;
class GlishRecord;

// <summary> A holder for Quantities to enable record conversions </summary>

// <use visibility=export>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="tQuantumHolder" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=RecordInterface>RecordInterface</linkto> class
//   <li> <linkto class=GlishRecord>GlishRecord</linkto> class
//   <li> <linkto class=Quantum>Quantity</linkto> class
// </prerequisite>
//
// <etymology>
// A Holder of general Quantities
// </etymology>
//
// <synopsis>
// This class can be used to handle a heterogeneous list of Quantities, and
// can handle torecord() and fromrecord() conversions.
// A QuantumHolder
// is created empty, from a Quantity (<src>Quantum<Double></src>), a
// <src>Quantum<Float></src> or a <src>Quantum<Int></src>.
//
// Checks on the contents can be made with functions like
// <src>isQuantity</src> and the contents can be obtained with
// functions like <src>asQuantity</src>. It is an error to try and
// retrieve a Quantum of the wrong type and doing so will generate an
// exception (AipsError).
// </synopsis>
//
// <example>
// <srcblock>
//	TableRecord rec;		// an empty record
//	Quantity x(12.5, "km/s");	// a Quantity
//	String error;			// an error message
//	if (!QuantumHolder(x).torecord(error, rec)) {  // make record
//		cout << error << endl;
//	};
// </srcblock>
// </example>
//
// <motivation>
// To make general conversions between Quantities and records, without knowing
// the actual Quantity being converted.
// </motivation>
//
// <todo asof="1998/04/14">
//   <li> possible change if GlishRecord derived from RecordInterface
// </todo>

class QuantumHolder : public RecordTransformable {

public:

//# Friends

//# Enumerations

//# Constructors
  // Creates an empty holder
  QuantumHolder();
  // Create from a Quantity (copy semantics)
  // <group>
  QuantumHolder(const Quantum<Double> &in);
  QuantumHolder(const Quantum<Float> &in);
  QuantumHolder(const Quantum<Int> &in);
  // </group>
  // Copy a holder (copy semantics)
  QuantumHolder(const QuantumHolder &other);
//# Destructor
  ~QuantumHolder();

//# Operators
  // Assignment (copy semantics)
  QuantumHolder &operator=(const QuantumHolder &other);
  // Get value (will have lifetime only as long as MeasHolder exists)
  const Quantum<Double> &operator()() const;

//# Member Functions
  // Check if it holds a Quantity
  // <group>
  Bool isQuantity() const;
  Bool isEmpty() const;
  // </group>

  // Get a Quantity from the holder (with lifetime as long 
  // as holder exists).
  // <thrown>
  // <li> AipsError if holder empty
  // </thrown>
  // <group>
  const Quantum<Double> &asQuantity() const;
  // </group>

  // Create a Quantity from a record
  // <group>
  virtual Bool fromRecord(String &error,
			  const RecordInterface &in);
  Bool fromRecord(String &error,
		  const GlishRecord &in);
  // </group>
  // Create a record from a Quantity
  // <group>
  virtual Bool toRecord(String &error, RecordInterface &out) const;
  Bool toRecord(String &error, GlishRecord &out) const;
  // </group>

private:

//# Data Members
  // Pointer to a Quantity
  PtrHolder<Quantum<Double> > hold;

};

#endif
