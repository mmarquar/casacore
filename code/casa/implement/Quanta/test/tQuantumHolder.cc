//# tQuantumHolder.cc: This program tests QuantumHolder
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes

#include <aips/aips.h>
#include <aips/Exceptions.h>
#include <aips/Measures/Quantum.h>
#include <trial/Measures/QuantumHolder.h>
#include <aips/Containers/Record.h>
#include <aips/Glish/GlishRecord.h>

main() {
  try {
    cout << "Test QuantumHolder  " << endl;
    cout << "----------------------------------------------------" << endl;

    String error;
    QuantumHolder q00;
    Quantity x00(12.5, "km/s");
    Quantum<Float> x01(30.3, "Jy/a");
    Quantum<Int> x02(2, "pc3/d");
    Quantity r00;
    Record y00;
    GlishRecord z00;
    
    cout << "Input quantity:              " << x00 << endl;
    if (QuantumHolder(x00).toRecord(error, y00)) {
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output quantity:      " << q00() << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };
    if (QuantumHolder(x00).toRecord(error, z00)) {
      if (q00.fromRecord(error, z00)) {
	cout <<"Glish output quantity:       " << q00() << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };

    cout << "Is quantity:                 " << q00.isQuantity() << endl;
    cout << "Is empty:                    " << q00.isEmpty() << endl;
    cout << "As quantity:                 " << q00.asQuantity() << endl;

    cout << "Input quantity:              " << x01 << endl;
    if (QuantumHolder(x01).toRecord(error, y00)) {
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output quantity:      " << q00() << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };
    cout << "Input quantity:              " << x02 << endl;
    if (QuantumHolder(x02).toRecord(error, y00)) {
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output quantity:      " << q00() << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };
    cout << "Error expected:" << endl;
    cout << "Input quantity:              " << x00 << endl;
    if (QuantumHolder(x00).toRecord(error, y00)) {
      y00.renameField("units", RecordFieldId("unit"));
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output quantity:      " << q00() << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };
  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } end_try;

  exit(0);
}
