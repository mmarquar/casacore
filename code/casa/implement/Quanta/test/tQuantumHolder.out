Test QuantumHolder  
----------------------------------------------------
Input quantity:              12.5 km/s
Record output quantity:      12.5 km/s
Glish output quantity:       12.5 km/s
Is quantity:                 1
Is empty:                    0
As quantity:                 12.5 km/s
Input quantity:              30.3 Jy/a
Record output quantity:      30.3 Jy/a
Input quantity:              2 pc3/d
Record output quantity:      2 pc3/d
Error expected:
Input quantity:              12.5 km/s
From error: Illegal Quantity record in QuantumHolder::fromRecord

